src/common/Throttle.h

```diff
@@ -362,15 +360,48 @@ class TokenBucketThrottle {
   CephContext *m_cct;
   Bucket m_throttle;
   uint64_t m_avg = 0;
+  uint64_t m_burst = 0;
   SafeTimer *m_timer;
   Mutex *m_timer_lock;
   FunctionContext *m_token_ctx = nullptr;
   list<Blocker> m_blockers;
   Mutex m_lock;
 
+  // minimum of the filling period.
+  uint64_t m_tick_min = 50;
+  // tokens filling period, its unit is millisecond.
+  uint64_t m_tick = 0;
+  /**
+   * These variables are used to calculate how many tokens need to be filled
+   * within each tick.
+   *
+   * In actual use, the tokens per tick may not be an integer, so take its
+   * integer part(m_tokens_per_tick) and spread the remainder(m_divide_left)
+   * into each tick to ensure that a set number of tokens can be transmitted
+   * per second.
+   *
+   * For example, we set the value of rate to be 950, means 950 iops(or bps).
+   * In this case, the actual period is 1000 / 950 = 1.052, too small for
+   * the SafeTimer. So the tick should set to be 50(m_tick_min), 20 ticks in
+   * one second.
+   * The tokens filled in bucket per tick is 950 / 20 = 47.5, not an integer.
+   * We take the 47 as m_tokens_per_tick, and there are 950 - (47 * 20) = 10
+   * tokens left. To fix this, we spread the remaining 10 tokens into 10 ticks.
+   *
+   * As a result, the tokens filled in one second will shown as this:
+   * tick    | 1| 2| 3| 4| 5| 6| 7| 8| 9|10|11|12|13|14|15|16|17|18|19|20|
+   * tokens  |48|48|48|48|48|48|48|48|48|48|47|47|47|47|47|47|47|47|47|47|
+   */
+  uint64_t m_ticks_per_second = 0;
+  uint64_t m_tokens_per_tick = 0;
+  uint64_t m_divide_left = 0;
```

Comment:
What a great idea! Thank you very much. :+1: 

But I think it will take more time to calculate (two floating multiplication and division calculations  within each tick). Or maybe it doesn't matter and I don't have to worry about it. :smile: 

I'll take your idea and think more about it. Thanks.

Reply:
:wink:  because I think there was too many if and else in m_divide_left way, then I give you a suggestion above, maybe you can find some better idea.

I think we should not worry too much on the overhead on calculation in this case.